lib/fs/spifs/spifs.c

```diff
@@ -0,0 +1,1146 @@
+
+    cursor_t cursor;
+    status = cursor_init(&cursor, spifs, spifs->toc_position, toc_page_addr,
+                         SPIFS_ENTRY_LENGTH);
+    if (status != NO_ERROR)
+        goto err;
+
+    toc_header_t *header = (toc_header_t*)cursor_get(&cursor);
+    spifs->num_entries = header->num_entries;
+    header = NULL;
+
+    // Create in-memory versions of metadata for files.
+    spifs_file_t *file;
+    for (size_t i = 0; i < spifs->num_entries; i++) {
+        status = cursor_advance(&cursor);
+        if (status != NO_ERROR)
+            goto err;
+
+        toc_file_t* file_entry = (toc_file_t*)cursor_get(&cursor);
+        if (file_entry->capacity == 0) {
+            continue;
+        }
+
+        file = malloc(sizeof(*file));
+        if (!file) {
+            status = ERR_NO_MEMORY;
+            goto err;
+        }
+
+        memcpy(&file->metadata, file_entry, SPIFS_ENTRY_LENGTH);
+
+        file->fs_handle = spifs;
+
+        list_add_tail(&spifs->files, &file->node);
+    }
+
+    if (!consistency_check(spifs)) {
+        status = ERR_BAD_STATE;
+        goto err;
+    }
+
+    *cookie = (fscookie *)spifs;
+
+    return NO_ERROR;
+
+err:
+    while ((file = list_remove_head_type(&spifs->files, spifs_file_t, node))) {
+        free(file);
+    }
+
+    free(spifs->page);
+    free(spifs);
+    return status;
+}
+
+static status_t spifs_unmount(fscookie *cookie)
+{
+    LTRACEF("cookie %p\n", cookie);
+
+    spifs_t *spifs = (spifs_t*)cookie;
+
+    mutex_acquire(&spifs->lock);
+
+    spifs_file_t* file;
+    while ((file = list_remove_head_type(&spifs->files, spifs_file_t, node))) {
+        free(file);
+    }
+
+    free(spifs->page);
+
+    mutex_release(&spifs->lock);
+
+    free(spifs);
+
+    return NO_ERROR;
+}
+
+static status_t spifs_create(fscookie *cookie, const char *name, filecookie **fcookie, uint64_t len)
+{
+    status_t status = NO_ERROR;
+
+    LTRACEF("cookie %p name '%s' filecookie %p len %llu\n", cookie, name, fcookie, len);
+
+    spifs_t *spifs = (spifs_t*)cookie;
+
+    // Strip leading fwd-slashes
+    name = trim_name(name);
+
+    // File system is flat, directories not supported.
+    if (strchr(name, '/'))
+        return ERR_NOT_SUPPORTED;
+
+    // Check that filename is not too long.
+    if (strnlen(name, MAX_FILENAME_LENGTH) == MAX_FILENAME_LENGTH)
+        return ERR_BAD_PATH;
+
+    mutex_acquire(&spifs->lock);
+
+    if (find_file(spifs, name)) {
+        status = ERR_ALREADY_EXISTS;
+        goto err;
+    }
+
+    // Is the ToC full? Have we reached the limit on the number of files?
+    size_t num_files_in_toc = list_length(&spifs->files);
+    DEBUG_ASSERT(num_files_in_toc <= spifs->num_entries);
+    if (num_files_in_toc >= spifs->num_entries) {
+        status = ERR_TOO_BIG;
+        goto err;
+    }
+
+    uint32_t capacity;
+    if (len == 0) {
+        capacity = spifs->page_size;
+    } else {
+        capacity = ROUNDUP(len, spifs->page_size);
```

Comment:
might want to check that the len arg isn't too big (> 4GB) which might truncate badly here